blocks/api/parser.js

```
@@ -147,10 +147,16 @@ export function getAttributesFromDeprecatedVersion( blockType, innerHTML, attrib
 			...omit( blockType, [ 'attributes', 'save', 'supports' ] ), // Parsing/Serialization properties
 			...blockType.deprecated[ i ],
 		};
-		const deprecatedBlockAttributes = getBlockAttributes( deprecatedBlockType, innerHTML, attributes );
-		const isValid = isValidBlock( innerHTML, deprecatedBlockType, deprecatedBlockAttributes );
-		if ( isValid ) {
-			return deprecatedBlockAttributes;
+
+		try {
+			const deprecatedBlockAttributes = getBlockAttributes( deprecatedBlockType, innerHTML, attributes );
+			const migratedBlockAttributes = deprecatedBlockType.migrate ? deprecatedBlockType.migrate( deprecatedBlockAttributes ) : deprecatedBlockAttributes;
+			const isValid = isValidBlock( innerHTML, deprecatedBlockType, deprecatedBlockAttributes );
+			if ( isValid ) {
+				return migratedBlockAttributes;
+			}
+		} catch ( error ) {
+			// ignore error, it means this deprecated version is invalid
```

The reasoning here is that the `migrate` function is meant to work properly only when it produces a valid block which means we should just ignore errors.

---

Where is the error being thrown?

---

There's no error nowhere. The idea is that `migrate` function could access non-existing attributes in invalid blocks. For example:

`migrate = ( oldAttributes ) => ( { newAttributeKey: oldAttributes.key1.key2 } )`

On an invalid block, this could throw an error since `oldAttributese.key1` could be undefined for instance while on valid blocks it can be mandatory for example. The idea is that the migrate function is meant to work only for valid blocks.

---

I think even the parsing could produce errors in theory

---

>On an invalid block, this could throw an error since `oldAttributese.key1` could be undefined for instance while on valid blocks it can be mandatory for example. The idea is that the migrate function is meant to work only for valid blocks.

I feel a bit uneasy about the fact that we anticipate that this will occur 😬